doc/03-Configuration.md
# Configuration

If Icinga Web has been installed but not yet set up, please visit Icinga Web and follow the web-based setup wizard.
For Icinga Web setups already running, log in to Icinga Web with a privileged user and follow the steps below to
configure Icinga DB Web:

If you have previously used the monitoring module, there is an option to [migrate](10-Migration.md) some settings.

## Database Configuration

Connection configuration for the database to which Icinga DB synchronizes monitoring data.

1. Create a new resource for the Icinga DB database via the `Configuration → Application → Resources` menu.

2. Configure the resource you just created as the database connection for the Icinga DB Web module using the
   `Configuration → Modules → icingadb → Database` menu.

This configuration is stored in `/etc/icingaweb2/modules/icingadb/config.ini`. The `resource`
option must match the name of the database resource you created for the Icinga DB database,
for example:

```
[icingadb]
resource = "icingadb"
```

If you use MySQL or MariaDB and notice poor performance for history queries, you can also
disable the query optimizer for these queries in the same section:

```
[icingadb]
...
disable_optimizer_for_history_queries = 1
```

This adds hints to disable the optimizer for history queries and may improve
performance on some MySQL/MariaDB versions.

## Redis® Configuration

Connection configuration for the Redis® server where Icinga 2 writes check results.
This data is used to display the latest state information in Icinga DB Web.

You can adjust the Redis® connection settings in the Web UI:

1. Configure the connection to the Redis® server through the `Configuration → Modules → icingadb → Redis` menu.

But you can also do this manually by editing the configuration file `/etc/icingaweb2/modules/icingadb/redis.ini`. The
basic structure of the file is as follows:

```ini
[redis]

[redis1]

[redis2]
```

The `redis` section is used for general options, `redis1` is used for the primary Redis® and `redis2` is used for the
secondary Redis® in case of high-availability setups. The general options are applied to all Redis® connections.

### Available settings

#### General Redis® Options

| Option | Description                                                                      |
|--------|----------------------------------------------------------------------------------|
| tls    | Enable TLS for the connection to Redis®. Set to `1` to enable, or `0` to disable |

#### Primary and Secondary Redis® Options

| Option   | Description                                                                |
|----------|----------------------------------------------------------------------------|
| host     | Hostname or IP address of the Redis® server.                               |
| port     | Port of the Redis® server. Defaults to `6380`.                             |
| database | Database number to use for the connection. Defaults to `0`.                |
| username | Username for the Redis® connection. If not set, no authentication is used. |
| password | Password for the Redis® connection. If not set, no authentication is used. |

!!! info

    If you are running a high-availability Icinga 2 setup,
    also configure the secondary master's Redis® connection details.
    Icinga DB Web then uses this connection if the primary one is not available.

## Command Transport Configuration

In order to acknowledge problems, force checks, schedule downtimes, etc.,
Icinga DB Web needs access to the Icinga 2 API.
For this you need an `ApiUser` object with at least the following permissions on the Icinga 2 side:

* `actions/*`
* `objects/query/*`
* `objects/modify/*`
* `status/query`

!!! tip

    For single-node setups it is recommended to manage API credentials in the `/etc/icinga2/conf.d/api-users.conf` file.
    If you are running a high-availability Icinga 2 setup, please manage the credentials in the master zone.

1. Please add the following Icinga 2 configuration and change the password accordingly:
   ```
   object ApiUser "icingadb-web" {
       password = "CHANGEME"
       permissions = [ "actions/*", "objects/modify/*", "objects/query/*", "status/query" ]
   }
   ```
2. Restart Icinga 2 for these changes to take effect.
3. Then configure a command transport for Icinga DB Web
   using the credentials you just created via the `Configuration → Modules → icingadb → Command Transports` menu.

But you can also do this manually by editing the configuration file
`/etc/icingaweb2/modules/icingadb/command_transports.ini`. Every command transport is defined in a separate section
with `[TRANSPORT_NAME]` as the section header.

### Available settings

| Option    | Description                                 |
|-----------|---------------------------------------------|
| transport | Type of command transport. Has to be `api`. |
| host      | Hostname or IP address of the Icinga 2 API. |
| port      | Port of the Icinga 2 API.                   |
| username  | Username for the Icinga 2 API connection.   |
| password  | Password for the Icinga 2 API connection.   |

!!! info

    If you are running a high-availability Icinga 2 setup,
    also configure the secondary master's API command transport.
    Icinga DB Web then uses this transport if the primary one is not available.

## General Configuration

You can adjust some default values of options users have while interacting with particular dialogs in the UI. (e.g. While acknowledging a problem)
These options can not be adjusted in the UI directly, but have to be set in the
configuration file `/etc/icingaweb2/modules/icingadb/config.ini`.

### Available Settings and defaults

| Option                            | Description                                                                                                                                                                                                                         | Default             |
|-----------------------------------|-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|---------------------|
| acknowledge_comment_text          | Sets the value for "Comment" in the Acknowledgement dialog                                                                                                                                                                          | ""                  |
| acknowledge_expire                | Toggles "Use Expire Time" in the Acknowledgement dialog.                                                                                                                                                                            | **0 (false)**       |
| acknowledge_expire_time           | Sets the value for "Expire Time" in the Acknowledgement dialog. It is calculated as now + this setting. Format is a [PHP Dateinterval](http://www.php.net/manual/en/dateinterval.construct.php).                                    | **1 hour (PT1H)**.  |
| acknowledge_notify                | Toggles "Send Notification" in the Acknowledgement dialog.                                                                                                                                                                          | **1 (true)**        |
| acknowledge_persistent            | Toggles "Persistent Comment" in the Acknowledgement dialog.                                                                                                                                                                         | **0 (false)**       |
| acknowledge_sticky                | Toggles "Sticky Acknowledgement" in the Acknowledgement dialog.                                                                                                                                                                     | **0 (false)**       |
| comment_expire                    | Toggles "Use Expire Time" in the Comment dialog.                                                                                                                                                                                    | **0 (false)**       |
| hostdowntime_comment_text         | Sets the value for "Comment" in the Host Downtime dialog                                                                                                                                                                            | ""                  |
| servicedowntime_comment_text      | Sets the value for "Comment" in the Service Downtime dialog.                                                                                                                                                                        | ""                  |
| comment_expire_time               | Sets the value for "Expire Time" in the Comment dialog. It is calculated as now + this setting. Format is a [PHP Dateinterval](http://www.php.net/manual/en/dateinterval.construct.php).                                            | **1 hour (PT1H)**   |
| custom_notification_forced        | Toggles "Forced" in the Custom Notification dialog.                                                                                                                                                                                 | **0 (false)**       |

CHANGELOG.md
# Icinga DB Web Changelog

Please make sure to always read our [Upgrading](https://icinga.com/docs/icinga-db-web/latest/doc/05-Upgrading/)
documentation before switching to a new version.

## 1.3.1 (2026-04-01)

All included changes can be found on the milestone: https://github.com/Icinga/icingadb-web/milestone/15?closed=1

### Notable Fixes

* Setups without the `monitoring` module no longer produce unnecessary integration errors (#1346)
* JSON exports with pagination now generate valid JSON without a leading comma (#1349)
* SLA reports scheduleded via the reporting module do not crash its daemon anymore (#1339)

## 1.3.0 (2025-11-19)

All included changes can be found on the milestone: https://github.com/Icinga/icingadb-web/milestone/10?closed=1

### Icinga Notifications

This release accompanies the release of Icinga Notifications v0.2.0. We have changed how notification sources are
configured, and now an integration into Icinga Web is mandatory. Icinga DB Web v1.3 will serve as the integration
for Icinga DB v1.5.

### New Features

* The *Services* tab of a host and the *History* tab of hosts and services are now filterable (#981, #1293)
* Exporting hosts and services to CSV or JSON by using the `?columns=…` query parameter now includes only the columns
  specified in the parameter in the results (#1011)

### Breaking Changes

* The restrictions `icingadb/denylist/variables` and `icingadb/protect/variables` from different roles are now
  merged into a single list, respectively. This means that variables denied in one role will not show up anymore
  if another role denies access to different variables. The same applies to `icingadb/protect/variables`, in which
  case variables protected in one role will now be protected even if another role protects different variables.
  This has been done to simplify the configuration and to get it more in line with how refusals work in Icinga Web.

### Maintenance

* The migration widget in the upper right has now been removed. We assume most users may be rather annoyed by it, and
  those who are still unsure when to migrate fully to Icinga DB are unlikely to be finally convinced by it anyway.
  Modules with support for Icinga DB will now default to use it if available.
* The upgrading documentation includes more details and other changes, make sure to read it carefully.

## 1.2.4 (2025-11-19)

This release includes fixes for the following issues:

* Commandtransport in cluster mode is broken #1292
* Redis config: Path "/nix/store" is not writable #1269
* Hide list selection-count and action list hint in minimal layout #1298

## 1.2.3 (2025-10-16)

**Notice:** This is a security release. It is recommended to upgrade _quickly_.

See the related CVE: https://github.com/Icinga/icingadb-web/security/advisories/GHSA-w57j-28jc-8429
Included changes can be found on the milestone: https://github.com/Icinga/icingadb-web/milestone/13?closed=1
And a detailed description about the most important ones on our blog: https://icinga.com/blog/releasing-icinga-2-v2-15-1-2-14-7-and-2-13-13-and-icinga-db-web-v1-2-3-and-1-1-4/

## 1.2.2 (2025-07-16)

**Notice:** This is a security release. It is recommended to upgrade _quickly_.

See the related CVE: https://github.com/Icinga/icingadb-web/security/advisories/GHSA-q2w7-mrx8-5473
Included changes can be found on the milestone: https://github.com/Icinga/icingadb-web/milestone/12?closed=1
And a detailed description about the most important ones on our blog: https://icinga.com/blog/releasing-icinga-web-2-v2-12-5-icinga-db-web-v1-2-2/

## 1.2.1 (2025-06-30)

This release [fixes an issue](https://github.com/Icinga/icingadb-web/pull/1229) with version comparison that caused
incorrect detection of the Icinga DB version. As a result, the health check in Icinga Web will now accurately reflect
the current Icinga DB version and no longer incorrectly indicate that an upgrade to version 1.4.0 is still required.

## 1.2.0 (2025-06-18)

Included changes can be found on the milestone: https://github.com/Icinga/icingadb-web/milestone/7?closed=1
And a detailed description about the most important ones on our blog: https://icinga.com/blog/icinga-dependency-views/

## 1.1.3 (2024-08-06)

Included changes can be found on the milestone: https://github.com/Icinga/icingadb-web/milestone/8?closed=1
And a detailed description about the most important ones on our blog: https://icinga.com/blog/2024/08/06/releasing-icinga-db-web-v1-1-3-and-a-ipl-security-release/

## 1.1.2 (2024-04-11)

Included changes can be found on the milestone: https://github.com/Icinga/icingadb-web/milestone/6?closed=1

configuration.php
<?php

// SPDX-FileCopyrightText: 2019 Icinga GmbH <https://icinga.com>
// SPDX-License-Identifier: GPL-3.0-or-later

namespace Icinga\Module\Icingadb {

    use Icinga\Application\Icinga;
    use Icinga\Authentication\Auth;
    use Icinga\Module\Icingadb\Web\Navigation\Renderer\HostProblemsBadge;
    use Icinga\Module\Icingadb\Web\Navigation\Renderer\ServiceProblemsBadge;
    use Icinga\Util\StringHelper;
    use RecursiveDirectoryIterator;
    use RecursiveIteratorIterator;

    /** @var \Icinga\Application\Modules\Module $this */

    $auth = Auth::getInstance();
    $authenticated = Icinga::app()->isWeb() && $auth->isAuthenticated();

    $this->provideSetupWizard('Icinga\Module\Icingadb\Setup\IcingaDbWizard');

    $this->providePermission(
        'icingadb/command/*',
        $this->translate('Allow all commands')
    );
    $this->providePermission(
        'icingadb/command/schedule-check',
        $this->translate('Allow to schedule host and service checks')
    );
    $this->providePermission(
        'icingadb/command/schedule-check/active-only',
        $this->translate('Allow to schedule host and service checks (Only on objects with active checks enabled)')
    );
    $this->providePermission(
        'icingadb/command/acknowledge-problem',
        $this->translate('Allow to acknowledge host and service problems')
    );
    $this->providePermission(
        'icingadb/command/remove-acknowledgement',
        $this->translate('Allow to remove problem acknowledgements')
    );
    $this->providePermission(
        'icingadb/command/comment/*',
        $this->translate('Allow to add and delete host and service comments')
    );
    $this->providePermission(
        'icingadb/command/comment/add',
        $this->translate('Allow to add host and service comments')
    );
    $this->providePermission(
        'icingadb/command/comment/delete',
        $this->translate('Allow to delete host and service comments')
    );
    $this->providePermission(
        'icingadb/command/downtime/*',
        $this->translate('Allow to schedule and delete host and service downtimes')
    );
    $this->providePermission(
        'icingadb/command/downtime/schedule',
        $this->translate('Allow to schedule host and service downtimes')
    );
    $this->providePermission(
        'icingadb/command/downtime/delete',
        $this->translate('Allow to delete host and service downtimes')
    );
    $this->providePermission(
        'icingadb/command/process-check-result',
        $this->translate('Allow to process host and service check results')
    );
    $this->providePermission(
        'icingadb/command/feature/instance',
        $this->translate('Allow to toggle instance-wide features')
    );
    $this->providePermission(
        'icingadb/command/feature/object/*',
        $this->translate('Allow to toggle all features on host and service objects')
    );
    $this->providePermission(
        'icingadb/command/feature/object/active-checks',
        $this->translate('Allow to toggle active checks on host and service objects')
    );
    $this->providePermission(
        'icingadb/command/feature/object/passive-checks',
        $this->translate('Allow to toggle passive checks on host and service objects')
    );
    $this->providePermission(
        'icingadb/command/feature/object/notifications',
        $this->translate('Allow to toggle notifications on host and service objects')
    );
    $this->providePermission(
        'icingadb/command/feature/object/event-handler',
        $this->translate('Allow to toggle event handlers on host and service objects')
    );
    $this->providePermission(
        'icingadb/command/feature/object/flap-detection',
        $this->translate('Allow to toggle flap detection on host and service objects')
    );
    $this->providePermission(
        'icingadb/command/send-custom-notification',
        $this->translate('Allow to send custom notifications for hosts and services')
    );

    $this->providePermission(
        'icingadb/object/show-source',
        $this->translate('Allow to view an object\'s source data. (May contain sensitive data!)')
    );

    $this->provideRestriction(
        'icingadb/filter/objects',
        $this->translate('Restrict access to the Icinga objects that match the filter')
    );

    $this->provideRestriction(
        'icingadb/filter/hosts',
        $this->translate('Restrict access to the Icinga hosts and services that match the filter')
    );

    $this->provideRestriction(
        'icingadb/filter/services',
        $this->translate('Restrict access to the Icinga services that match the filter')
    );

    $this->provideRestriction(
        'icingadb/denylist/routes',
        $this->translate('Prevent access to routes that are part of the list')
    );

    $this->provideRestriction(
        'icingadb/denylist/variables',
        $this->translate('Hide custom variables of Icinga objects that are part of the list')
    );

    $this->provideRestriction(
        'icingadb/protect/variables',
        $this->translate('Obfuscate custom variable values of Icinga objects that are part of the list')
    );

    if (! $this::exists('monitoring') || ($authenticated && ! $auth->getUser()->can('module/monitoring'))) {
        $routeDenylist = [];
        if ($authenticated && ! $auth->getUser()->isUnrestricted()) {
            // The empty array is for PHP pre 7.4, older versions require at least a single param for array_merge
            $routeDenylist = array_flip(array_merge([], ...array_map(function ($restriction) {
                return StringHelper::trimSplit($restriction);
            }, $auth->getRestrictions('icingadb/denylist/routes'))));
        }

        /*
        * Available navigation items
        */
        $this->provideNavigationItem('icingadb-host-action', $this->translate('Host Action'));
        $this->provideNavigationItem('icingadb-service-action', $this->translate('Service Action'));

        /**
         * Search urls
         */
        $this->provideSearchUrl(
            $this->translate('Tactical Overview'),
            'icingadb/tactical',
            100
        );
        $this->provideSearchUrl(
            $this->translate('Hosts'),
            'icingadb/hosts?sort=host.state.severity&limit=10',
            99
        );
        $this->provideSearchUrl(
            $this->translate('Services on Hosts'),
            'icingadb/services?sort=service.state.severity&limit=10&_hostFilterOnly',
            98
        );
        $this->provideSearchUrl(
            $this->translate('Services'),
            'icingadb/services?sort=service.state.severity&limit=10',
            97
        );

        if (! array_key_exists('hostgroups', $routeDenylist)) {
            $this->provideSearchUrl(
                $this->translate('Hostgroups'),
                'icingadb/hostgroups?limit=10',
                96
            );
        }

        if (! array_key_exists('servicegroups', $routeDenylist)) {
            $this->provideSearchUrl(
                $this->translate('Servicegroups'),
                'icingadb/servicegroups?limit=10',
                95
            );
        }

        /**
         * Current Incidents
         */
        $dashboard = $this->dashboard(N_('Current Incidents'), ['priority' => 50]);
        $dashboard->add(
            N_('Service Problems'),
            'icingadb/services?service.state.is_problem=y'
            . '&view=minimal&limit=32&sort=service.state.severity desc',
            100
        );
        $dashboard->add(
            N_('Recently Recovered Services'),
            'icingadb/services?service.state.soft_state=0'
            . '&view=minimal&limit=32&sort=service.state.last_state_change desc',
            110
        );
        $dashboard->add(
            N_('Host Problems'),
            'icingadb/hosts?host.state.is_problem=y'
            . '&view=minimal&limit=32&sort=host.state.severity desc',
            120
        );

        /**
         * Overdue
         */
        $dashboard = $this->dashboard(N_('Overdue'), ['priority' => 70]);
        $dashboard->add(
            N_('Late Host Check Results'),
            'icingadb/hosts?host.state.is_overdue=y'
            . '&view=minimal&limit=15&sort=host.state.severity desc',
            100
        );
        $dashboard->add(
            N_('Late Service Check Results'),
            'icingadb/services?service.state.is_overdue=y'
            . '&view=minimal&limit=15&sort=service.state.severity desc',
            110
        );
        $dashboard->add(
            N_('Acknowledgements Active For At Least Three Days'),
            'icingadb/comments?comment.entry_type=ack&comment.entry_time<-3 days'
            . '&view=minimal&limit=15&sort=comment.entry_time',
            120
        );
        $dashboard->add(
            N_('Downtimes Active For At Least Three Days'),
            'icingadb/downtimes?downtime.is_in_effect=y&downtime.scheduled_start_time<-3 days'
            . '&view=minimal&limit=15&sort=downtime.start_time',
            130
        );

        /**
         * Muted
         */
        $dashboard = $this->dashboard(N_('Muted'), ['priority' => 80]);
        $dashboard->add(
            N_('Disabled Service Notifications'),
            'icingadb/services?service.notifications_enabled=n'
            . '&view=minimal&limit=15&sort=service.state.severity desc',
            100
        );
        $dashboard->add(
            N_('Disabled Host Notifications'),
            'icingadb/hosts?host.notifications_enabled=n'
            . '&view=minimal&limit=15&sort=host.state.severity desc',
            110
        );
        $dashboard->add(
            N_('Disabled Service Checks'),
            'icingadb/services?service.active_checks_enabled=n'
            . '&view=minimal&limit=15&sort=service.state.last_state_change',
            120
        );
        $dashboard->add(
            N_('Disabled Host Checks'),
            'icingadb/hosts?host.active_checks_enabled=n'
            . '&view=minimal&limit=15&sort=host.state.last_state_change',
            130
        );
        $dashboard->add(
            N_('Acknowledged Problem Services'),
            'icingadb/services?service.state.is_acknowledged!=n&service.state.is_problem=y'
            . '&view=minimal&limit=15&sort=service.state.severity desc',
            140
        );
        $dashboard->add(
            N_('Acknowledged Problem Hosts'),
            'icingadb/hosts?host.state.is_acknowledged!=n&host.state.is_problem=y'
            . '&view=minimal&limit=15&sort=host.state.severity desc',
            150
        );

        /**
         * @var \Icinga\Application\Modules\Module $this
         *
         * Problems section in case monitoring is disabled
         */
        $problemSection = $this->menuSection(N_('Problems'), [
            'renderer' => array(
                'TotalProblemsBadge',
                'state' => 'critical'
            ),
            'icon'     => 'attention-circled',
            'priority' => 20
        ]);
        $problemSection->add(N_('Host Problems'), [
            'renderer'    => (new HostProblemsBadge())->disableLink(),
            'icon'        => 'server',
            'description' => $this->translate('List current host problems'),
            'url'         => 'icingadb/hosts?host.state.is_problem=y'
                . '&sort=host.state.severity desc',
            'priority'    => 50
        ]);
        $problemSection->add(N_('Service Problems'), [
            'renderer'    => (new ServiceProblemsBadge())->disableLink(),
            'icon'        => 'cog',
            'description' => $this->translate('List current service problems'),
            'url'         => 'icingadb/services?service.state.is_problem=y'
                . '&sort=service.state.severity desc',
            'priority'    => 60
        ]);
        $problemSection->add(N_('Service Grid'), [
            'icon'        => 'cogs',
            'description' => $this->translate('Display service problems as grid'),
            'url'         => 'icingadb/services/grid?problems',
            'priority'    => 70
        ]);

        $problemSection->add(N_('Current Downtimes'), [
            'description' => $this->translate('List current downtimes'),
            'url'         => 'icingadb/downtimes?downtime.is_in_effect=y',
            'priority'    => 80,
            'icon'        => 'plug'
        ]);

        /**
         * @var \Icinga\Application\Modules\Module $this
         *
         * Overview section in case monitoring is disabled
         */
        $overviewSection = $this->menuSection('Overview', [
            'icon'     => 'binoculars',
            'priority' => 30
        ]);

        $overviewSection->add(N_('Tactical Overview'), [
            'url'         => 'icingadb/tactical',
            'description' => $this->translate('Open tactical overview'),
            'priority'    => 40,
            'icon'        => 'chart-pie'
        ]);
        $overviewSection->add(N_('Hosts'), [
            'priority'    => 50,
            'description' => $this->translate('List hosts'),
            'url'         => 'icingadb/hosts',
            'icon'        => 'server'
        ]);
        $overviewSection->add(N_('Services'), [
            'priority'    => 60,
            'description' => $this->translate('List services'),
            'url'         => 'icingadb/services',
            'icon'        => 'cog'
        ]);

        if (! array_key_exists('hostgroups', $routeDenylist)) {
            $overviewSection->add(N_('Host Groups'), [
                'description' => $this->translate('List host groups'),
                'url'         => 'icingadb/hostgroups',
                'priority'    => 70,
                'icon'        => 'network-wired'
            ]);
        }

        if (! array_key_exists('servicegroups', $routeDenylist)) {
            $overviewSection->add(N_('Service Groups'), [
                'description' => $this->translate('List service groups'),
                'url'         => 'icingadb/servicegroups',
                'priority'    => 80,
                'icon'        => 'cogs'
            ]);
        }

        if (! array_key_exists('contactgroups', $routeDenylist)) {
            $overviewSection->add(N_('Contact Groups'), [
                'description' => $this->translate('List contact groups'),
                'url'         => 'icingadb/contactgroups',
                'priority'    => 90,
                'icon'        => 'users'
            ]);
        }

        if (! array_key_exists('contacts', $routeDenylist)) {
            $overviewSection->add(N_('Contacts'), [
                'description' => $this->translate('List contacts'),
                'url'         => 'icingadb/contacts',
                'priority'    => 100,
                'icon'        => 'user-friends'
            ]);
        }

        $overviewSection->add(N_('Comments'), [
            'url'         => 'icingadb/comments',
            'description' => $this->translate('List comments'),
            'priority'    => 110,
            'icon'        => 'comments'
        ]);
        $overviewSection->add(N_('Downtimes'), [
            'url'         => 'icingadb/downtimes',
            'description' => $this->translate('List downtimes'),
            'priority'    => 120,
            'icon'        => 'plug'
        ]);

        /**
         * @var \Icinga\Application\Modules\Module $this
         *
         * History section in case monitoring is disabled
         */

        $section = $this->menuSection(N_('History'), array(
            'icon'     => 'history',
            'priority' => 90
        ));
        $section->add(N_('Notifications'), array(
            'icon'        => 'bell',
            'description' => $this->translate('List notifications'),
            'priority'    => 20,
            'url'         => 'icingadb/notifications',
        ));
        $section->add(N_('Event Overview'), array(
            'icon'        => 'history',
            'description' => $this->translate('Open event overview'),
            'priority'    => 30,
            'url'         => 'icingadb/history'
        ));
    } else {
        /*
        * Available navigation items
        */
        $this->provideNavigationItem(
            'icingadb-host-action',
            $this->translate('Host Action') . ' (Icinga DB)'
        );
        $this->provideNavigationItem(
            'icingadb-service-action',
            $this->translate('Service Action') . ' (Icinga DB)'
        );

        /** @var \Icinga\Application\Modules\Module $this */
        $section = $this->menuSection('Icinga DB', [
            'icon'     => 'database',
            'priority' => 30
        ]);

        $section->add(N_('Tactical Overview'), [
            'url'           => 'icingadb/tactical',
            'priority'      => 10,
            'description'   => $this->translate('Open tactical overview'),
            'icon'          => 'chart-pie'
        ]);

        $section->add(N_('Hosts'), [
            'priority'      => 20,
            'description'   => $this->translate('List hosts'),
            'renderer'      => 'HostProblemsBadge',
            'url'           => 'icingadb/hosts',
            'icon'          => 'server'
        ]);
        $section->add(N_('Services'), [
            'priority'      => 30,
            'description'   => $this->translate('List services'),
            'renderer'      => 'ServiceProblemsBadge',
            'url'           => 'icingadb/services',
            'icon'          => 'cog'
        ]);
        $section->add(N_('Service Grid'), [
            'icon'        => 'cog',
            'description' => $this->translate('Display service problems as grid'),
            'url'         => 'icingadb/services/grid?problems',
            'priority'    => 40
        ]);

        $routeDenylist = [];
        if ($authenticated && ! $auth->getUser()->isUnrestricted()) {
            // The empty array is for PHP pre 7.4, older versions require at least a single param for array_merge
            $routeDenylist = array_flip(array_merge([], ...array_map(function ($restriction) {
                return StringHelper::trimSplit($restriction);
            }, $auth->getRestrictions('icingadb/denylist/routes'))));
        }

        if (! array_key_exists('hostgroups', $routeDenylist)) {
            $section->add(N_('Host Groups'), [
                'url'           => 'icingadb/hostgroups',
                'priority'      => 50,
                'description'   => $this->translate('List host groups'),
                'icon'          => 'network-wired'
            ]);
        }

        if (! array_key_exists('servicegroups', $routeDenylist)) {
            $section->add(N_('Service Groups'), [
                'url'           => 'icingadb/servicegroups',
                'priority'      => 60,
                'description'   => $this->translate('List service groups'),
                'icon'          => 'cogs'
            ]);
        }

        if (! array_key_exists('contactgroups', $routeDenylist)) {
            $section->add(N_('Contact Groups'), [
                'url'           => 'icingadb/contactgroups',
                'priority'      => 70,
                'description'   => $this->translate('List contact groups'),
                'icon'          => 'users'
            ]);
        }

        if (! array_key_exists('contacts', $routeDenylist)) {
            $section->add(N_('Contacts'), [
                'url'           => 'icingadb/contacts',
                'priority'      => 80,
                'description'   => $this->translate('List contacts'),
                'icon'          => 'user-friends'
            ]);
        }

        $section->add(N_('Comments'), [
            'url'           => 'icingadb/comments',
            'priority'      => 90,
            'description'   => $this->translate('List comments'),
            'icon'          => 'comments'
        ]);
        $section->add(N_('Downtimes'), [
            'url'           => 'icingadb/downtimes',
            'priority'      => 100,
            'description'   => $this->translate('List downtimes'),
            'icon'          => 'plug'
        ]);
        $section->add(N_('Notifications'), [
            'url'           => 'icingadb/notifications',
            'priority'      => 110,
            'description'   => $this->translate('List notifications'),
            'icon'          => 'bell'
        ]);
        $section->add(N_('History'), [
            'url'           => 'icingadb/history',
            'priority'      => 120,
            'description'   => $this->translate('List history'),
            'icon'          => 'history'
        ]);
    }

    $this->provideConfigTab('database', [
        'label' => t('Database'),
        'title' => t('Configure the database backend'),
        'url'   => 'config/database'
    ]);
    $this->provideConfigTab('redis', [
        'label' => t('Redis'),
        'title' => t('Configure the Redis connections'),
        'url'   => 'config/redis'
    ]);
    $this->provideConfigTab('command-transports', [
        'label' => t('Command Transports'),
        'title' => t('Configure command transports'),
        'url'   => 'command-transport'
    ]);

    $cssDirectory = $this->getCssDir();
    $cssFiles = new RecursiveIteratorIterator(new RecursiveDirectoryIterator(
        $cssDirectory,
        RecursiveDirectoryIterator::CURRENT_AS_PATHNAME | RecursiveDirectoryIterator::SKIP_DOTS
    ));
    foreach ($cssFiles as $path) {
        $this->provideCssFile(ltrim(substr($path, strlen($cssDirectory)), DIRECTORY_SEPARATOR));
    }

    $this->provideJsFile('progress-bar.js');
    $this->provideJsFile('timestamp-toggle.js');
}
